packages/flutter/lib/src/widgets/overflow_bar.dart

```diff
@@ -0,0 +1,487 @@
+///
+/// {@tool dartpad --template=stateless_widget_scaffold_center}
+///
+/// This example defines a simple approximation of a dialog
+/// layout, where the layout of the dialog's action buttons are
+/// defined by an [OverflowBar].  The content is wrapped in a
+/// [SingleChildScrollView], so that if overflow occurs, the
+/// action buttons will be still be accessible by scrolling,
+/// no matter how much vertical space is available.
+///
+/// ```dart
+/// Widget build(BuildContext context) {
+///   return Container(
+///     alignment: Alignment.center,
+///     padding: EdgeInsets.all(16),
+///     color: Colors.black.withOpacity(0.15),
+///     child: Material(
+///       color: Colors.white,
+///       elevation: 24,
+///       shape: RoundedRectangleBorder(
+///         borderRadius: BorderRadius.all(Radius.circular(4))
+///       ),
+///       child: Padding(
+///         padding: EdgeInsets.all(8),
+///         child: SingleChildScrollView(
+///           child: Column(
+///             mainAxisSize: MainAxisSize.min,
+///             crossAxisAlignment: CrossAxisAlignment.stretch,
+///             children: <Widget>[
+///               Container(height: 128, child: Placeholder()),
+///               Align(
+///                 alignment: AlignmentDirectional.centerEnd,
+///                 child: OverflowBar(
+///                   spacing: 8,
+///                   overflowAlignment: OverflowBarAlignment.end,
+///                   children: <Widget>[
+///                     TextButton(child: Text('Cancel'), onPressed: () { }),
+///                     TextButton(child: Text('Really Really Cancel'), onPressed: () { }),
+///                     OutlinedButton(child: Text('OK'), onPressed: () { }),
+///                   ],
+///                 ),
+///               ),
+///             ],
+///           ),
+///         ),
+///       ),
+///     ),
+///   );
+/// }
+/// ```
+/// {@end-tool}
+class OverflowBar extends MultiChildRenderObjectWidget {
+  /// Constructs an OverflowBar.
+  ///
+  /// The [spacing], [overflowSpacing], [overflowAlignment],
+  /// [overflowDirection], and [clipBehavior] parameters must not be
+  /// null.
+  OverflowBar({
+    Key key,
+    this.spacing = 0.0,
+    this.overflowSpacing = 0.0,
+    this.overflowAlignment = OverflowBarAlignment.start,
+    this.overflowDirection = VerticalDirection.down,
+    this.textDirection,
+    this.clipBehavior = Clip.none,
+    List<Widget> children = const <Widget>[],
+  }) : assert(spacing != null),
+       assert(overflowSpacing != null),
+       assert(overflowAlignment != null),
+       assert(overflowDirection != null),
+       assert(clipBehavior != null),
+       super(key: key, children: children);
+
+  /// The width of the gap between [children] for the default
+  /// horizontal layout.
+  ///
+  /// If the horizontal layout overflows, then [overflowSpacing] is
+  /// used instead.
+  ///
+  /// Defaults to 0.0.
+  final double spacing;
```

Comment:
I wonder if we should make this default match the material spec for button spacing? Or is that assuming too much? It would just be nice to just wrap an array of buttons with an `OverflowBar` with no params and have it match the old `ButtonBar` layout by default.

Reply:
For a dialog you'd also need to end-align the OverflowBar and somewhere you need to ensure that the OverflowBar can be scrolled into view.

Using 0 for the spacing seemed appropriate for the widgets library since it's probably the wrong place to have an opinion about spacing.

Reply:
Ah, I missed that it was part of the widgets library 😄. Totally makes sense for it to be there and that the default should be 0.